File: jianzhi_offer(剑指offer)/数据流中的中位数.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <cstring>
#include <queue>
#include <sstream>
using namespace std;

// 2018.9.6
class Solution {
private:
    priority_queue<int, vector<int>, std::greater<int> > min_q;
    priority_queue<int, vector<int>, std::less<int> > max_q;

public:
    void Insert(int num)
    {
        if(max_q.empty() || num > max_q.top()){
            min_q.push(num);
        } else{
            max_q.push(num);
        } if(max_q.size()+2 == min_q.size()){  // 2020.5.9 → min_q.size()-max_q.size() == 2 → min_q.size()-max_q.size() > 1 这时就不能通过，应该跟size()函数的unsigned int有关
            max_q.push(min_q.top());
            min_q.pop();
        } else if(max_q.size() == min_q.size()+1){
            min_q.push(max_q.top());
            max_q.pop();
        }
    }

    double GetMedian()
    {
        if(max_q.size() == min_q.size()){
            return (max_q.top()+min_q.top())/2.0;
        } else{
            return min_q.top();
        }
    }

};

int main(){
    Solution solution = Solution();
    priority_queue<int, vector<int>, std::greater<int> > min_queue;
    for(int i = 10; i >= 0; --i){
        min_queue.push(i);
    } cout << min_queue.top() << endl;
    return 0;
}

File: jianzhi_offer(剑指offer)/数据流中的中位数.java
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.Queue;

public class Solution {
    private Queue<Integer> minH = new PriorityQueue<Integer>();
    private Queue<Integer> maxH = new PriorityQueue<Integer>(
        new Comparator<Integer>(){
            @Override
            public int compare(Integer x, Integer y){
                return y-x;
            }
        });
    public void Insert(Integer num) {
        if(maxH.isEmpty() || num <= maxH.peek()){
            maxH.add(num);
        } else{
            minH.add(num);
        }
        if(minH.size()+2 == maxH.size()){
            minH.offer(maxH.poll());
        }
        else if(maxH.size() < minH.size()){
            maxH.offer(minH.poll());
        }
    }

    public Double GetMedian() {
        if(minH.size() == maxH.size()){
            return (minH.peek()+maxH.peek())/2.0;
        } return maxH.peek()*1.0;
    }
}
